py4cast/models/vision/unetrpp.py

```diff
@@ -243,16 +253,31 @@ def __init__(
         # keys and values from HWD-dimension to P-dimension
         self.EF = nn.Parameter(init_(torch.zeros(input_size, proj_size)))
 
+        if self.use_scaled_dot_product_SA:
+            # we need an extra set of weights for the scaled dot product
+            self.FE = nn.Parameter(init_(torch.zeros(proj_size, input_size)))
+        else:
+            # the original unetr++ uses two parameters for scaling the dot products
+            self.temperature2 = nn.Parameter(torch.ones(num_heads, 1, 1))
+
+        if not self.use_scaled_dot_product_CA:
+            self.temperature = nn.Parameter(torch.ones(num_heads, 1, 1))
+
         self.attn_drop = nn.Dropout(channel_attn_drop)
         self.attn_drop_2 = nn.Dropout(spatial_attn_drop)
 
     def forward(self, x):
         B, N, C = x.shape
 
         qkvv = self.qkvv(x).reshape(B, N, 4, self.num_heads, C // self.num_heads)
+
+        # Matrix index, Batch, Head, Dimensions, Features
         qkvv = qkvv.permute(2, 0, 3, 1, 4)
+
+        # Batch, Head, Dimensions, Features
         q_shared, k_shared, v_CA, v_SA = qkvv[0], qkvv[1], qkvv[2], qkvv[3]
 
+        # Batch, Head, Features, Dimensions
         q_shared = q_shared.transpose(-2, -1)
         k_shared = k_shared.transpose(-2, -1)
```

Comment:
Could we remove the transpose() calls and add them when doing the permute(): `qkvv = qkvv.permute(2, 0, 3, 4, 1)` instead the original one ?

Reply:
We could. For this PR I choose to stick as close as possible to the original implementation while adding the sdpa to minimize risks and facilitate comparison. I think we probably should have a got at simplifying/optmising all these in a dedicated PR (also because the current PR comes with 5 days of training to validate the results)